Implement replaceOrCreate and replaceAttributes

Amir-61 opened the pull request:
Please See Spike task: https://github.com/strongloop-internal/scrum-loopback/issues/621

##### Added functionality:

*Initial implementation for replaceAttribute
*Initial implementation for replaceOrCreate

##### Other notes:

* These changes are related to the changes in the pull requests under `loopback-connector` and `loopback-datasource-juggler`
* It seems test cases cannot be run on windows machines; I need to dig into it.

Connect to https://github.com/strongloop-internal/scrum-loopback/issues/621

Amir-61 commented on lib/mongodb.js:
```diff
@@ -483,6 +482,121 @@ MongoDB.prototype.updateOrCreate = function updateOrCreate(model, data, options,
+ *
+ * @param data
+ * @returns {*}
+ */
```
@bajtos: Again probably it makes more sense to have an argument in `parseUpdateData` for `replace` and if value `true` is passed, we unset the missing fields by `parsedData.$unset[key] = "";`

bajtos replied:
Yes please, don't duplicate code if possible.

I am proposing to remove extended operators support from `replace`, as the intention is different here.

bajtos commented on lib/mongodb.js:
```diff
@@ -428,7 +428,6 @@ MongoDB.prototype.parseUpdateData = function(model, data, options) {
   } else {
     parsedData.$set = data;
   }
-
```
Unrelated whitespace change, please remove.

bajtos commented on lib/mongodb.js:
```diff
@@ -802,6 +916,41 @@ MongoDB.prototype.count = function count(model, where, options, callback) {
+    object && idName !== '_id' && delete object._id;
+    cb && cb(err, object);
+  });
+};
```
The implementation of `replaceAttributes` and `replaceOrCreate` look mostly identical to me. Is there a way how to extract shared code to a single shared method?

bajtos commented:
@raymondfeng @superkhau you are more familiar with MongoDB connector than I am, could you PTAL?

bajtos commented on lib/mongodb.js:
```diff
@@ -483,6 +482,121 @@ MongoDB.prototype.updateOrCreate = function updateOrCreate(model, data, options,
+      parsedData.$unset[key] = "";
+    }
+  }
+  return parsedData;
```
I think the solution based on `{ $set, $unset }` will not work correctly when the data contains any extra properties not defined in the model. I believe we need to use [Replace the document](https://docs.mongodb.org/manual/tutorial/modify-documents/#replace-the-document) variant instead.

> To replace the entire content of a document except for the _id field, pass an entirely new document as the second argument to update().

Amir-61 replied:
Oh yea that's right!

bajtos commented on lib/mongodb.js:
```diff
@@ -483,6 +483,32 @@ MongoDB.prototype.updateOrCreate = function updateOrCreate(model, data, options,
+    if (err)
+      return callback && callback(err);
+    if (self.debug)
+      debug('replaceOrCreate.callback', model, {_id: oid}, data, err, info);
```
`err` is always falsy here. I think it's better to move logging on L504-505 above the error check on L502-503?

bajtos commented on lib/mongodb.js:
```diff
@@ -483,6 +483,32 @@ MongoDB.prototype.updateOrCreate = function updateOrCreate(model, data, options,
+    debug('replaceOrCreate', model, data);
+  }
+  var id = self.getIdValue(model, data);
+  var idName = self.idName(model);
```
AFAICT `idName` is not used, can it be removed?

Amir-61 replied:
I used it to delete from the data object `delete data[idName];`

superkhau replied:
this.idName(...)

bajtos commented on lib/mongodb.js:
```diff
@@ -802,6 +828,28 @@ MongoDB.prototype.count = function count(model, where, options, callback) {
+    if (err)
+      return cb && cb(err);
+    if (self.debug)
+      debug('replaceAttributes.callback', model, {_id: oid}, data, err, info);
```
`err` is always falsy here. I think it's better to move logging on L504-505 above the error check on L502-503?

bajtos commented:
@Amir-61 the code looks much better now! 

Could you please refactor the code to reuse as much of the shared code as possible?

Example:

```js
MongoDB.prototype.replaceOrCreate = function replaceOrCreate(model, data, options, callback) {
  var id = self.getIdValue(model, data);
  var oid = self.coerceId(model, id);
  delete data[idName];
  this._replaceWithOptions(model, oid, data, {upsert: true}, cb);
}

MongoDB.prototype.replaceAttributes = function replaceAttrs(model, id, data, options, cb) {
  var oid = self.coerceId(model, id);
  this._replaceWithOptions(model, oid, data, {upsert: false}, cb);
}
```

Amir-61 commented:
@superkhau Please review; also it is worth mentioning this patch works with my changes from `loopback-datasource-juggle` and the test cases work only when I have my changes from loopback-datasource-juggler in place; so it cannot be merged yet.

Please let me know your feedback Thanks a lot!

superkhau commented on lib/mongodb.js:
```diff
@@ -483,6 +483,25 @@ MongoDB.prototype.updateOrCreate = function updateOrCreate(model, data, options,
+ * @param {Function} [cb] The callback function
+ */
+MongoDB.prototype.replaceOrCreate = function (model, data, options, cb) {
+  var self = this;
```
Is `var self = this` even need? I don't see any case in this function that requires it.

superkhau commented on lib/mongodb.js:
```diff
@@ -483,6 +483,25 @@ MongoDB.prototype.updateOrCreate = function updateOrCreate(model, data, options,
+  if (self.debug) {
+    debug('replaceOrCreate', model, data);
+  }
+  var id = self.getIdValue(model, data);
```
this.getIdValue(...)

superkhau commented on lib/mongodb.js:
```diff
@@ -483,6 +483,25 @@ MongoDB.prototype.updateOrCreate = function updateOrCreate(model, data, options,
+  }
+  var id = self.getIdValue(model, data);
+  var idName = self.idName(model);
+  var oid = self.coerceId(model, id);
```
this.coerceId(...)

superkhau commented on lib/mongodb.js:
```diff
@@ -483,6 +483,25 @@ MongoDB.prototype.updateOrCreate = function updateOrCreate(model, data, options,
+ * @param {Object} data The model instance data
+ * @param {Function} [cb] The callback function
+ */
+MongoDB.prototype.replaceOrCreate = function (model, data, options, cb) {
```
Remove space after `function`. Should look like `function(model, data, ...)`.

superkhau commented on lib/mongodb.js:
```diff
@@ -802,6 +821,48 @@ MongoDB.prototype.count = function count(model, where, options, callback) {
+ * @param {Object} data The model data
+ * @param {Function} [cb] The callback function
+ */
+MongoDB.prototype.replaceAttributes = function replaceAttrs(model, id, data, options, cb) {
```
Remove `replaceAttrs`. Should look like `... = function(model, id, data, ...)`.

superkhau commented on lib/mongodb.js:
```diff
@@ -802,6 +821,48 @@ MongoDB.prototype.count = function count(model, where, options, callback) {
+ * @param {Function} [cb] The callback function
+ */
+MongoDB.prototype.replaceAttributes = function replaceAttrs(model, id, data, options, cb) {
+  var self = this;
```
`self = this` not required.

superkhau commented on lib/mongodb.js:
```diff
@@ -802,6 +821,48 @@ MongoDB.prototype.count = function count(model, where, options, callback) {
+ */
+MongoDB.prototype.replaceAttributes = function replaceAttrs(model, id, data, options, cb) {
+  var self = this;
+  if (self.debug) {
```
this.debug

superkhau commented on lib/mongodb.js:
```diff
@@ -802,6 +821,48 @@ MongoDB.prototype.count = function count(model, where, options, callback) {
+  if (self.debug) {
+    debug('replaceAttributes', model, id, data);
+  }
+  var oid = self.coerceId(model, id);
```
this.coerceId

superkhau commented on lib/mongodb.js:
```diff
@@ -802,6 +821,48 @@ MongoDB.prototype.count = function count(model, where, options, callback) {
+ * Replace properties for the model instance data
+ * @param {String} model The model name
+ * @param {*} id The instance id
+ * @param {Object} data The model data
```
missing `options` description

superkhau commented on lib/mongodb.js:
```diff
@@ -483,6 +483,25 @@ MongoDB.prototype.updateOrCreate = function updateOrCreate(model, data, options,
+ * Replace if the model instance exists with the same id or create a new instance
+ *
+ * @param {String} model The model name
+ * @param {Object} data The model instance data
```
missing `options` description

superkhau replied:
The model instance's data/attributes

superkhau commented on lib/mongodb.js:
```diff
@@ -802,6 +821,48 @@ MongoDB.prototype.count = function count(model, where, options, callback) {
+/**
+ * Update a model instance with id
+ * @param {String} model The model name
+ * @param {Object} where The search criteria which are key value pairs
```
wrong desc. `where` should be id or vice versa.

superkhau commented on lib/mongodb.js:
```diff
@@ -802,6 +821,48 @@ MongoDB.prototype.count = function count(model, where, options, callback) {
+ * @callback {Function} [cb] Callback function
+ */
+MongoDB.prototype.replaceWithOptions = function(model, id, data, options, cb) {
+  var self =this;
```
space after =

superkhau commented on lib/mongodb.js:
```diff
@@ -802,6 +821,48 @@ MongoDB.prototype.count = function count(model, where, options, callback) {
+ */
+MongoDB.prototype.replaceWithOptions = function(model, id, data, options, cb) {
+  var self =this;
+  this.execute(model, 'update', {_id: id}, data, options, function (err, info) {
```
remove the extra space between function and (

superkhau commented on lib/mongodb.js:
```diff
@@ -802,6 +821,48 @@ MongoDB.prototype.count = function count(model, where, options, callback) {
+  var self =this;
+  this.execute(model, 'update', {_id: id}, data, options, function (err, info) {
+    if (self.debug)
+      debug('updateWithOptions.callback', model, {_id: id}, data, err, info);
```
add empty line here to separate logic

superkhau commented on lib/mongodb.js:
```diff
@@ -802,6 +821,48 @@ MongoDB.prototype.count = function count(model, where, options, callback) {
+    if (self.debug)
+      debug('updateWithOptions.callback', model, {_id: id}, data, err, info);
+    if (err)
+      return cb && cb(err);
```
and here

superkhau commented on test/mongodb.test.js:
```diff
@@ -1072,6 +1072,109 @@ describe('mongodb connector', function () {
     });
   });
 
+  describe('replaceOrCreate', function () {
```
remove space btwn fn and (

superkhau commented on test/mongodb.test.js:
```diff
@@ -1072,6 +1072,109 @@ describe('mongodb connector', function () {
   });
 
+  describe('replaceOrCreate', function () {
+    it('replaceOrCreate: should create a model instance even if it does already exist', function (done) {
```
remove `replaceOrCreate: `, mocha will print it out anyways

superkhau commented on test/mongodb.test.js:
```diff
@@ -1072,6 +1072,109 @@ describe('mongodb connector', function () {
+      });
+    });
+
+    it('replaceOrCreate: should replace a model instance if it the passing key already exists', function (done) {
```
same here

superkhau commented on test/mongodb.test.js:
```diff
@@ -1072,6 +1072,109 @@ describe('mongodb connector', function () {
+      });
+    });
+
+    it('replaceOrCreate: should remove extra properties not defined in the model', function (done) {
```
and here

superkhau commented on test/mongodb.test.js:
```diff
@@ -1072,6 +1072,109 @@ describe('mongodb connector', function () {
+  })  
+  
+  describe('replaceAttributes', function () {
+    it('replaceAttributes: should replace a model instance if it the passing key already exists', function (done) {
```
remove `replaceAttributes: `

superkhau commented on test/mongodb.test.js:
```diff
@@ -1072,6 +1072,109 @@ describe('mongodb connector', function () {
+      });
+    });
+
+    it('replaceAttributes: should remove extra properties not defined in the model', function (done) {
```
here too

superkhau commented on test/mongodb.test.js:
```diff
@@ -1072,6 +1072,109 @@ describe('mongodb connector', function () {
+
+  });
+
+  describe('replaceAttribute', function () {
```
here

superkhau commented on test/mongodb.test.js:
```diff
@@ -1072,6 +1072,109 @@ describe('mongodb connector', function () {
+      });
+    });
+
+    it('replaceAttribute: should remove extra properties not defined in the model', function (done) {
```
here

superkhau commented on test/mongodb.test.js:
```diff
@@ -1072,6 +1072,109 @@ describe('mongodb connector', function () {
+          updatedProduct.name.should.be.equal('newFoo');
+          should.exist(updatedProduct.id);
+          Product.findById(product.id, function (err, data) {
+            data.name.should.be.equal('newFoo');
```
remove space btwn fn and (

superkhau commented on test/mongodb.test.js:
```diff
@@ -1072,6 +1072,109 @@ describe('mongodb connector', function () {
+      Product.create({name: 'foo', price: 100, 'bar': 'baz'}, function (err, product) {
+        Product.replaceOrCreate({id: product.id, name: 'newFoo'}, function (err, updatedProduct) {
+          should.not.exist(updatedProduct.bar);
+          Product.findById(product.id, function (err, data) {
```
remove space fn (

superkhau commented on test/mongodb.test.js:
```diff
@@ -1072,6 +1072,109 @@ describe('mongodb connector', function () {
+    });
+  })  
+  
+  describe('replaceAttributes', function () {
```
here

Amir-61 commented:
@superkhau  Thanks so much for the review. I applied all of your syntax suggestions...

Thanks again for the feedback :-)

superkhau commented on lib/mongodb.js:
```diff
@@ -483,6 +483,25 @@ MongoDB.prototype.updateOrCreate = function updateOrCreate(model, data, options,
+ * @param {Function} [cb] The callback function
+ */
+MongoDB.prototype.replaceOrCreate = function(model, data, options, cb) {
+  if (this.debug) {
```
{} not required for single line statements

```
if (this.debug)
  debug('replaceOrCreate', model, data);

var id = ...
```

Same goes for all the other debugs.

superkhau commented on lib/mongodb.js:
```diff
@@ -483,6 +483,25 @@ MongoDB.prototype.updateOrCreate = function updateOrCreate(model, data, options,
+    debug('replaceOrCreate', model, data);
+  }
+  var id = this.getIdValue(model, data);
+  var idName = this.idName(model);
```
I would move this down one line to put it close to where you're using it.

superkhau commented on lib/mongodb.js:
```diff
@@ -802,6 +821,49 @@ MongoDB.prototype.count = function count(model, where, options, callback) {
+MongoDB.prototype.replaceAttributes = function replaceAttrs(model, id, data, options, cb) {
+  if (this.debug) {
+    debug('replaceAttributes', model, id, data);
+  }
```
debug curly

superkhau commented on lib/mongodb.js:
```diff
@@ -802,6 +821,49 @@ MongoDB.prototype.count = function count(model, where, options, callback) {
+ * @param {String} model The model name
+ * @param {Object} id The id of the model instance
+ * @param {Object} data The property/value pairs to be updated or inserted if {upsert: true} is passed as options
+ * @param {Object} options The options you wnat to pass for update, e.g, {upsert: true}
```
wnat = want

superkhau commented on lib/mongodb.js:
```diff
@@ -483,6 +483,25 @@ MongoDB.prototype.updateOrCreate = function updateOrCreate(model, data, options,
 };
 
 /**
+ * Replace if the model instance exists with the same id or create a new instance
```
Replace model instance if it exists or create a new one if it doesn't.

superkhau commented on lib/mongodb.js:
```diff
@@ -483,6 +483,25 @@ MongoDB.prototype.updateOrCreate = function updateOrCreate(model, data, options,
 /**
+ * Replace if the model instance exists with the same id or create a new instance
+ *
+ * @param {String} model The model name
```
The name of the model

superkhau commented on lib/mongodb.js:
```diff
@@ -483,6 +483,25 @@ MongoDB.prototype.updateOrCreate = function updateOrCreate(model, data, options,
+ * @param {Object} options The options object
+ * @param {Function} [cb] The callback function
+ */
+MongoDB.prototype.replaceOrCreate = function(model, data, options, cb) {
```
model -> modelName is more descriptive

Amir-61 replied:
Hey Simon, I applied all of your feedback except changing `model` to `modelName` in many places you mentioned this change makes the code inconsistent with the rest of the code; in other words, in other methods others are using `model` as well and refactoring the other code written by others seems huge changes...

superkhau replied:
Sounds good. The rename can be done in a separate commit/PR.

superkhau commented on lib/mongodb.js:
```diff
@@ -483,6 +483,25 @@ MongoDB.prototype.updateOrCreate = function updateOrCreate(model, data, options,
+ *
+ * @param {String} model The model name
+ * @param {Object} data The model instance data
+ * @param {Object} options The options object
```
What options are available?

superkhau commented on lib/mongodb.js:
```diff
@@ -483,6 +483,25 @@ MongoDB.prototype.updateOrCreate = function updateOrCreate(model, data, options,
+
+/**
  * Delete a model instance by id
  * @param {String} model The model name
```
Same with model -> modelName

superkhau commented on lib/mongodb.js:
```diff
@@ -802,6 +821,49 @@ MongoDB.prototype.count = function count(model, where, options, callback) {
 
 /**
+ * Replace properties for the model instance data
+ * @param {String} model The model name
```
and here

superkhau commented on lib/mongodb.js:
```diff
@@ -802,6 +821,49 @@ MongoDB.prototype.count = function count(model, where, options, callback) {
+ * @param {Object} options The options object
+ * @param {Function} [cb] The callback function
+ */
+MongoDB.prototype.replaceAttributes = function replaceAttrs(model, id, data, options, cb) {
```
and here

raymondfeng replied:
I'm not sure about the method name - `replaceAttributes` hints that only the attrs are replaced. Can we use `replace` instead?

Amir-61 replied:
That's a good question; please see this [comment](https://github.com/strongloop/loopback-datasource-juggler/pull/788#issuecomment-164428158)

superkhau replied:
I'm +1 with @raymondfeng on the method name, why not just `replace` instead.

superkhau commented on lib/mongodb.js:
```diff
@@ -802,6 +821,49 @@ MongoDB.prototype.count = function count(model, where, options, callback) {
+
+/**
+ * Update a model instance with id
+ * @param {String} model The model name
```
and here

superkhau commented on lib/mongodb.js:
```diff
@@ -802,6 +821,49 @@ MongoDB.prototype.count = function count(model, where, options, callback) {
+ * @param {Object} options The options you wnat to pass for update, e.g, {upsert: true}
+ * @callback {Function} [cb] Callback function
+ */
+MongoDB.prototype.replaceWithOptions = function(model, id, data, options, cb) {
```
and here

superkhau commented on test/mongodb.test.js:
```diff
@@ -1072,6 +1072,111 @@ describe('mongodb connector', function () {
   });
 
+  describe('replaceOrCreate', function() {
+    it('should create a model instance even if it does already exist2', function(done) {
```
exist2 -> exist

superkhau commented on test/mongodb.test.js:
```diff
@@ -1072,6 +1072,111 @@ describe('mongodb connector', function () {
+      });
+    });
+
+    it('should replace a model instance if it the passing key already exists', function(done) {
```
if it ... -> if the passing

superkhau commented on test/mongodb.test.js:
```diff
@@ -1072,6 +1072,111 @@ describe('mongodb connector', function () {
+      });
+    });
+
+    it('should remove extra properties not defined in the model', function(done) {
```
should remove extraneous properties that are not defined in the model

superkhau commented on test/mongodb.test.js:
```diff
@@ -1072,6 +1072,111 @@ describe('mongodb connector', function () {
+  })  
+  
+  describe('replaceAttributes', function() {
+    it('should replace a model instance if it the passing key already exists', function(done) {
```
replace the model instance if the provided key already exists

superkhau commented on test/mongodb.test.js:
```diff
@@ -1072,6 +1072,111 @@ describe('mongodb connector', function () {
+      });
+    });
+
+    it('should remove extra properties not defined in the model', function(done) {
```
should remove extraneous properties that are not defined in the model

superkhau commented on test/mongodb.test.js:
```diff
@@ -1072,6 +1072,111 @@ describe('mongodb connector', function () {
+      });
+    });
+
+    it('should remove extra properties not defined in the model', function(done) {
```
should remove extraneous properties that are not defined in the model

superkhau commented on test/mongodb.test.js:
```diff
@@ -1072,6 +1072,111 @@ describe('mongodb connector', function () {
+  });
+
+  describe('replaceAttribute', function() {
+    it('should replace a model instance if it the passing key already exists', function(done) {
```
replace the model instance if the provided key already exists

bajtos commented on test/mongodb.test.js:
```diff
@@ -1072,6 +1072,111 @@ describe('mongodb connector', function () {
+  describe('replaceOrCreate', function() {
+    it('should create a model instance even if it already exists', function(done) {
+      Product.replaceOrCreate({name: 'newFoo'}, function(err, updatedProduct) {
+        should.not.exist(err);
```
This produces unhelpful error messages. Please always use `if (err) return done(err)` in unit-tests, even if that's not consistent with the legacy code around.

Amir-61 replied:
I'll fix them...

bajtos commented on test/mongodb.test.js:
```diff
@@ -1072,6 +1072,111 @@ describe('mongodb connector', function () {
+
+  });
+
+  describe('replaceAttribute', function() {
```
I believe `replaceAttribute` has been removed?

Amir-61 replied:
If I understood your question correctly; you are asking for implementation of `replaceAttribute` in `loopback-connector-mongodb`?

But that is handled in `loopback-datasource-juggler` like the following:

```
DataAccessObject.prototype.replaceAttribute = function replaceAttribute(name, value, options, cb) {
  var data = {};
  data[name] = value;
  return this.replaceAttributes(data, options, cb);
};
```

Please correct me if I did not understand your question or I'm mistaken

bajtos commented:
Let's wait with landing this until the PR for juggler is landed first.

Amir-61 commented:
Hi @bajtos,

> Let's wait with landing this until the PR for juggler is landed first: strongloop/loopback-datasource-juggler#788

That makes sense; however implementation of `replaceOrCreate` was part of my current sprint, so I applied your feedback and although in test cases I followed the same style of code as the rest of the code, I prefer to use `async.series` to get rid of callback hells; could you please let me know if you are OK with using `async` module in `mongodb.test.js` file? Or alternatively I can modularize the code to get rid of the callback hells...

Thanks alot!

Amir-61 commented:
Hi @bajtos,

Actually now I modularized most of the test cases...

Thanks!

bajtos commented:
> I prefer to use async.series or promises to get rid of callback hells; could you please let me know if you are OK with using async module in mongodb.test.js file

Sure, using `async` is a good idea! Promises would be even better, but we are not there yet...

raymondfeng commented:
I am out of the office until 01/04/2016.

Happy holidays!


Note: This is an automated response to your message  "Re:
[loopback-connector-mongodb] Implement replaceOrCreate and
replaceAttributes (#202)" sent on 12/21/15 2:25:04 PM.

This is the only notification you will receive while this person is away.

bajtos commented:
@Amir-61 I am reassigning this PR back to you, so that it stays out of my list of PRs waiting for my review. Please assign it back to me once you have removed `updateAttribute` (singular) and the PR is ready for another round of review.

Amir-61 commented:
> Please assign it back to me once you have removed updateAttribute (singular) and the PR is ready for another round of review.

@bajtos I removed `replaceAttribute`

> I prefer to use async.series or promises to get rid of callback hells; could you please let me know if you are OK with using async module in mongodb.test.js file

I guess based on you comments from my the other PR I should not overuse `async` as it adds a bit overhead, but by any chance if you still think I should use them especially in my test scripts, please let me know and I will re-factore them to use `async`...

Thanks!

Amir-61 commented:
Hi @superkhau,

Most of the issues reported by you are code format issues(e.g removing space between `function` keyword and `(`); so I was wondering why we don't have a linter like [eslint](https://github.com/eslint/eslint) integrated with this project? Whenever I push my code they all pass locally; in other words, if I get this code format issues as a part of test, I can fix them before sending a pull request and we can save lots of time...

CC: @bajtos

Thanks!

superkhau commented:
>Most of the issues reported by you are code format issues (e.g. removing space between function keyword and (); so I was wondering why we don't have a linter like eslint integrated with this project? Whenever I push my code they all pass locally; in other words, if I get these code format issues as a part of test, I can fix them before sending a pull request and we can save lots of time...

We currently use jshint in other projects and we in the process of upgrading to [eslint](https://github.com/strongloop/loopback/issues/1749). I've bumped the issue up to major and assigned it to myself to make code reviews easier. ;) Feel free to add eslint to this project in another PR.

superkhau commented:
>I guess based on you comments from my the other PR I should not overuse async as it adds a bit overhead, but by any chance if you still think I should use them especially in my test scripts, please let me know and I will re-factore them to use async...

Correctness then clarity. Premature optimization is evil.

bajtos commented:
> We currently use jshint in other projects and we in the process of upgrading to eslint. I've bumped the issue up to major and assigned it to myself to make code reviews easier. ;) Feel free to add eslint to this project in another PR.
> I don't know why jshint was forgotten for this project specifically too. ;) I'm going to go over all the projects once I do a spike on the eslint options we want.

Note that jshint does not check code formatting, you need jscs for that.

> > I guess based on you comments from my the other PR I should not overuse async as it adds a bit overhead, but by any chance if you still think I should use them especially in my test scripts, please let me know and I will re-factore them to use async...

> Correctness then clarity. Premature optimization is evil. IMO, you should use async for clarity unless you can prove (via a benchmark) that you actually need that part of the code optimized. Otherwise, you're making it harder for me to read/maintain. ;)

Both premature optimization and premature pessimizations are evil. For simple cases where the callbacks nest at most 2-3 levels, it's IMO better to avoid async, especially in production code.

bajtos commented on lib/mongodb.js:
```diff
@@ -802,6 +821,49 @@ MongoDB.prototype.count = function count(model, where, options, callback) {
+      debug('updateWithOptions.callback', model, {_id: id}, data, err, info);
+    
+    if (err)
+      return cb && cb(err);
```
Nitpick (feel free to ignore): when both `if` and the next line are short enough, we usually join them to a single line:

```js
if (err) return cb && cb(err);
```

The same comment applies to most of your tests.

bajtos commented on test/mongodb.test.js:
```diff
@@ -1072,6 +1072,122 @@ describe('mongodb connector', function () {
+        should.exist(updatedProduct.id);
+        verifyData(updatedProduct.id);
+      });
+       function verifyData(id) {
```
You spacing is a bit off here, looks like `function verifyData` has one more space in front?

bajtos commented:
@raymondfeng I'm handing over the review to you, as we agreed in SlackChat. 

I think it we should wait with landing this patch until https://github.com/strongloop/loopback-datasource-juggler/pull/788 is finished first.

superkhau commented:
>Both premature optimization and premature pessimizations are evil. For simple cases where the callbacks nest at most 2-3 levels, it's IMO better to avoid async, especially in production code. 

I can live with this, for simple cases, its better not to depend on yet another dependency like async in prod code unless you deem the extra complication is necessary for clarity.

Amir-61 commented:
> @raymondfeng I'm handing over the review to you, as we agreed in SlackChat.
Thanks @bajtos for following up :-) I applied your feedback; @raymondfeng Please review this patch. Thanks a lot!

> I think it we should wait with landing this patch until strongloop/loopback-datasource-juggler#788 is finished first.
Agree with this... Hopefully I get back to my spike task (strongloop/loopback-datasource-juggler#788) in the next sprint... With that being said if I can get a confirmation from @raymondfeng for this patch in this spring, it will be perfect, since I had the implementation of the `replaceAttributes` and `replaceOrCreate` for Mongo DB in my current sprint...

Thanks!

Amir-61 commented:
Update: Got LGTM from @raymondfeng and this patch will be merged after I move back to my spike task: https://github.com/strongloop/loopback-datasource-juggler/pull/788 and get it done which needs to be merged before this patch.

Thanks @raymondfeng!

raymondfeng commented:
@Amir-61 I prefer to rename `replaceAttributes` with `replace` before you land the PR. See similar comments in https://github.com/strongloop/loopback-datasource-juggler/pull/788

superkhau commented:
>@Amir-61 I prefer to rename replaceAttributes with replace before you land the PR. See similar comments in strongloop/loopback-datasource-juggler#788

+1

Amir-61 commented on lib/mongodb.js:
```diff
@@ -802,6 +820,47 @@ MongoDB.prototype.count = function count(model, where, options, callback) {
+    } else {
+      result = undefined;
+    }
+    cb && cb(err, result, cbInfo);
```
@raymondfeng: I had forgot to add this `cbInfo` which is used for `replaceOrCreate` to see if it is a `isNewInstance` or not in juggler for `after save` operation hook...
CC:  @bajtos

Amir-61 commented:
>@Amir-61 I prefer to rename replaceAttributes with replace before you land the PR. See similar comments in strongloop/loopback-datasource-juggler#788


@raymondfeng Thanks! I just renamed it also I added one more [change](https://github.com/strongloop/loopback-connector-mongodb/pull/202#discussion-diff-49133017)

raymondfeng commented:
LGTM

Amir-61 commented:
@slnode test please

Amir-61 commented:
@slnode test please

Amir-61 commented:
@slnode test please

rmg commented:
The failures on CI are interesting.. it is passing on all Amazon and failing on all Ubuntu.. that points very strongly at a problem caused by the version of MongoDB (2.4 on Ubuntu, 2.6 on Amazon).

Here's the failing tests:

```
  1) mongodb imported features manipulation replaceOrCreate works without options on create (promise variant):
     TypeError: Cannot read property 'id' of null
      at node_modules/loopback-datasource-juggler/test/manipulation.test.js:717:14
      at lib$es6$promise$$internal$$tryCatch (node_modules/mongodb/node_modules/es6-promise/dist/es6-promise.js:319:16)
      at lib$es6$promise$$internal$$invokeCallback (node_modules/mongodb/node_modules/es6-promise/dist/es6-promise.js:331:17)
      at lib$es6$promise$$internal$$publish (node_modules/mongodb/node_modules/es6-promise/dist/es6-promise.js:302:11)
      at Object.lib$es6$promise$asap$$flush [as _onImmediate] (node_modules/mongodb/node_modules/es6-promise/dist/es6-promise.js:113:9)

  2) mongodb imported features manipulation replaceOrCreate works with options on create (promise variant):
     TypeError: Cannot read property 'id' of null
      at node_modules/loopback-datasource-juggler/test/manipulation.test.js:739:14
      at lib$es6$promise$$internal$$tryCatch (node_modules/mongodb/node_modules/es6-promise/dist/es6-promise.js:319:16)
      at lib$es6$promise$$internal$$invokeCallback (node_modules/mongodb/node_modules/es6-promise/dist/es6-promise.js:331:17)
      at lib$es6$promise$$internal$$publish (node_modules/mongodb/node_modules/es6-promise/dist/es6-promise.js:302:11)
      at Object.lib$es6$promise$asap$$flush [as _onImmediate] (node_modules/mongodb/node_modules/es6-promise/dist/es6-promise.js:113:9)

  3) mongodb imported features manipulation replaceOrCreate works without options on create (callback variant):
     Uncaught TypeError: Cannot read property 'id' of null
      at node_modules/loopback-datasource-juggler/test/manipulation.test.js:876:14
      at node_modules/loopback-datasource-juggler/lib/dao.js:1811:62
      at node_modules/loopback-datasource-juggler/lib/dao.js:1743:9
      at Object.async.each (node_modules/loopback-datasource-juggler/node_modules/async/lib/async.js:153:20)
      at allCb (node_modules/loopback-datasource-juggler/lib/dao.js:1679:13)
      at lib/mongodb.js:9:43639
      at handleCallback (node_modules/mongodb/lib/utils.js:96:12)
      at node_modules/mongodb/lib/cursor.js:841:16
      at handleCallback (node_modules/mongodb/node_modules/mongodb-core/lib/cursor.js:244:5)
      at setCursorNotified (node_modules/mongodb/node_modules/mongodb-core/lib/cursor.js:563:3)
      at node_modules/mongodb/node_modules/mongodb-core/lib/cursor.js:673:16
      at queryCallback (node_modules/mongodb/node_modules/mongodb-core/lib/cursor.js:322:5)
      at Callbacks.emit (node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:95:3)
      at null.messageHandler (node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:249:23)
      at Socket.<anonymous> (node_modules/mongodb/node_modules/mongodb-core/lib/connection/connection.js:265:22)
      at Socket.<anonymous> (_stream_readable.js:765:14)
      at emitReadable_ (_stream_readable.js:427:10)
      at emitReadable (_stream_readable.js:423:5)
      at readableAddChunk (_stream_readable.js:166:9)
      at Socket.Readable.push (_stream_readable.js:128:10)
      at TCP.onread (net.js:529:21)

  4) mongodb imported features manipulation replaceOrCreate works with options on create (callback variant):
     Uncaught TypeError: Cannot read property 'id' of null
      at node_modules/loopback-datasource-juggler/test/manipulation.test.js:895:14
      at node_modules/loopback-datasource-juggler/lib/dao.js:1811:62
      at node_modules/loopback-datasource-juggler/lib/dao.js:1743:9
      at Object.async.each (node_modules/loopback-datasource-juggler/node_modules/async/lib/async.js:153:20)
      at allCb (node_modules/loopback-datasource-juggler/lib/dao.js:1679:13)
      at lib/mongodb.js:9:43639
      at handleCallback (node_modules/mongodb/lib/utils.js:96:12)
      at node_modules/mongodb/lib/cursor.js:841:16
      at handleCallback (node_modules/mongodb/node_modules/mongodb-core/lib/cursor.js:244:5)
      at setCursorNotified (node_modules/mongodb/node_modules/mongodb-core/lib/cursor.js:563:3)
      at node_modules/mongodb/node_modules/mongodb-core/lib/cursor.js:673:16
      at queryCallback (node_modules/mongodb/node_modules/mongodb-core/lib/cursor.js:322:5)
      at Callbacks.emit (node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:95:3)
      at null.messageHandler (node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:249:23)
      at Socket.<anonymous> (node_modules/mongodb/node_modules/mongodb-core/lib/connection/connection.js:265:22)
      at Socket.<anonymous> (_stream_readable.js:765:14)
      at emitReadable_ (_stream_readable.js:427:10)
      at emitReadable (_stream_readable.js:423:5)
      at readableAddChunk (_stream_readable.js:166:9)
      at Socket.Readable.push (_stream_readable.js:128:10)
      at TCP.onread (net.js:529:21)

  5) Persistence hooks PersistedModel.replaceOrCreate triggers `loaded` hook on create:

      Uncaught AssertionError: expected Object {
  Model: '[ModelCtor TestModel]',
  data: Object { id: 'new-id', name: 'a name' },
  hookState: Object { test: true },
  isNewInstance: false,
  options: Object {}
} to equal Object {
  Model: '[ModelCtor TestModel]',
  data: Object { id: 'new-id', name: 'a name' },
  hookState: Object { test: true },
  isNewInstance: true,
  options: Object {}
} (at isNewInstance, A has false and B has true)
      + expected - actual

         }
         "hookState": {
           "test": true
         }
      -  "isNewInstance": false
      +  "isNewInstance": true
         "options": {}
       }
      
      at Assertion.fail (node_modules/should/lib/assertion.js:92:17)
      at Assertion.Object.defineProperty.value (node_modules/should/lib/assertion.js:164:19)
      at node_modules/loopback-datasource-juggler/test/persistence-hooks.suite.js:2403:39
      at node_modules/loopback-datasource-juggler/lib/dao.js:778:19
      at doNotify (node_modules/loopback-datasource-juggler/lib/observer.js:93:49)
      at doNotify (node_modules/loopback-datasource-juggler/lib/observer.js:93:49)
      at Function.ObserverMixin._notifyBaseObservers (node_modules/loopback-datasource-juggler/lib/observer.js:116:5)
      at Function.ObserverMixin.notifyObserversOf (node_modules/loopback-datasource-juggler/lib/observer.js:91:8)
      at Function.ObserverMixin._notifyBaseObservers (node_modules/loopback-datasource-juggler/lib/observer.js:114:15)
      at Function.ObserverMixin.notifyObserversOf (node_modules/loopback-datasource-juggler/lib/observer.js:91:8)
      at node_modules/loopback-datasource-juggler/lib/dao.js:775:23
      at node_modules/loopback-datasource-juggler/lib/observer.js:106:23
      at node_modules/loopback-datasource-juggler/node_modules/async/lib/async.js:189:25
      at node_modules/loopback-datasource-juggler/test/persistence-hooks.suite.js:2951:18
      at notifySingleObserver (node_modules/loopback-datasource-juggler/lib/observer.js:98:22)
      at iterate (node_modules/loopback-datasource-juggler/node_modules/async/lib/async.js:181:13)
      at Object.async.eachSeries (node_modules/loopback-datasource-juggler/node_modules/async/lib/async.js:197:9)
      at doNotify (node_modules/loopback-datasource-juggler/lib/observer.js:95:11)
      at doNotify (node_modules/loopback-datasource-juggler/lib/observer.js:93:49)
      at Function.ObserverMixin._notifyBaseObservers (node_modules/loopback-datasource-juggler/lib/observer.js:116:5)
      at Function.ObserverMixin.notifyObserversOf (node_modules/loopback-datasource-juggler/lib/observer.js:91:8)
      at Function.ObserverMixin._notifyBaseObservers (node_modules/loopback-datasource-juggler/lib/observer.js:114:15)
      at Function.ObserverMixin.notifyObserversOf (node_modules/loopback-datasource-juggler/lib/observer.js:91:8)
      at done (node_modules/loopback-datasource-juggler/lib/dao.js:754:19)
      at lib/mongodb.js:9:49037
      at node_modules/loopback-datasource-juggler/lib/observer.js:166:22
      at doNotify (node_modules/loopback-datasource-juggler/lib/observer.js:93:49)
      at MongoDB.ObserverMixin._notifyBaseObservers (node_modules/loopback-datasource-juggler/lib/observer.js:116:5)
      at MongoDB.ObserverMixin.notifyObserversOf (node_modules/loopback-datasource-juggler/lib/observer.js:91:8)
      at cbForWork (node_modules/loopback-datasource-juggler/lib/observer.js:156:14)
      at args.(anonymous function) (lib/mongodb.js:9:11402)
      at handleCallback (node_modules/mongodb/lib/utils.js:96:12)
      at node_modules/mongodb/lib/collection.js:987:5
      at node_modules/mongodb/node_modules/mongodb-core/lib/wireprotocol/2_4_support.js:429:9

  6) Persistence hooks PersistedModel.replaceOrCreate triggers `after save` hook on create:

      Uncaught AssertionError: expected Object {
  Model: '[ModelCtor TestModel]',
  hookState: Object { test: true },
  instance: Object { extra: null, id: 'new-id', name: 'a name' },
  isNewInstance: false,
  options: Object {}
} to equal Object {
  Model: '[ModelCtor TestModel]',
  hookState: Object { test: true },
  instance: Object { extra: null, id: 'new-id', name: 'a name' },
  isNewInstance: true,
  options: Object {}
} (at isNewInstance, A has false and B has true)
      + expected - actual

           "extra": [null]
           "id": "new-id"
           "name": "a name"
         }
      -  "isNewInstance": false
      +  "isNewInstance": true
         "options": {}
       }
      
      at Assertion.fail (node_modules/should/lib/assertion.js:92:17)
      at Assertion.Object.defineProperty.value (node_modules/should/lib/assertion.js:164:19)
      at node_modules/loopback-datasource-juggler/test/persistence-hooks.suite.js:2498:39
      at node_modules/loopback-datasource-juggler/lib/dao.js:778:19
      at node_modules/loopback-datasource-juggler/lib/observer.js:106:23
      at node_modules/loopback-datasource-juggler/node_modules/async/lib/async.js:189:25
      at node_modules/loopback-datasource-juggler/test/persistence-hooks.suite.js:2951:18
      at notifySingleObserver (node_modules/loopback-datasource-juggler/lib/observer.js:98:22)
      at iterate (node_modules/loopback-datasource-juggler/node_modules/async/lib/async.js:181:13)
      at Object.async.eachSeries (node_modules/loopback-datasource-juggler/node_modules/async/lib/async.js:197:9)
      at doNotify (node_modules/loopback-datasource-juggler/lib/observer.js:95:11)
      at doNotify (node_modules/loopback-datasource-juggler/lib/observer.js:93:49)
      at Function.ObserverMixin._notifyBaseObservers (node_modules/loopback-datasource-juggler/lib/observer.js:116:5)
      at Function.ObserverMixin.notifyObserversOf (node_modules/loopback-datasource-juggler/lib/observer.js:91:8)
      at Function.ObserverMixin._notifyBaseObservers (node_modules/loopback-datasource-juggler/lib/observer.js:114:15)
      at Function.ObserverMixin.notifyObserversOf (node_modules/loopback-datasource-juggler/lib/observer.js:91:8)
      at node_modules/loopback-datasource-juggler/lib/dao.js:775:23
      at doNotify (node_modules/loopback-datasource-juggler/lib/observer.js:93:49)
      at doNotify (node_modules/loopback-datasource-juggler/lib/observer.js:93:49)
      at Function.ObserverMixin._notifyBaseObservers (node_modules/loopback-datasource-juggler/lib/observer.js:116:5)
      at Function.ObserverMixin.notifyObserversOf (node_modules/loopback-datasource-juggler/lib/observer.js:91:8)
      at Function.ObserverMixin._notifyBaseObservers (node_modules/loopback-datasource-juggler/lib/observer.js:114:15)
      at Function.ObserverMixin.notifyObserversOf (node_modules/loopback-datasource-juggler/lib/observer.js:91:8)
      at done (node_modules/loopback-datasource-juggler/lib/dao.js:754:19)
      at lib/mongodb.js:9:49037
      at node_modules/loopback-datasource-juggler/lib/observer.js:166:22
      at doNotify (node_modules/loopback-datasource-juggler/lib/observer.js:93:49)
      at MongoDB.ObserverMixin._notifyBaseObservers (node_modules/loopback-datasource-juggler/lib/observer.js:116:5)
      at MongoDB.ObserverMixin.notifyObserversOf (node_modules/loopback-datasource-juggler/lib/observer.js:91:8)
      at cbForWork (node_modules/loopback-datasource-juggler/lib/observer.js:156:14)
      at args.(anonymous function) (lib/mongodb.js:9:11402)
      at handleCallback (node_modules/mongodb/lib/utils.js:96:12)
      at node_modules/mongodb/lib/collection.js:987:5
      at node_modules/mongodb/node_modules/mongodb-core/lib/wireprotocol/2_4_support.js:429:9
```

Amir-61 commented:
Thanks @rmg a lot :-)

Amir-61 commented:
The only method I introduced in this patch is [update method](https://docs.mongodb.org/manual/tutorial/modify-documents/#replace-the-document) to replace a document. I agree with what @rmg says it probably behaves differently in different versions of mongodb. I'm using version 3.2.1 for mongodb ATM; I will try `2.4` which failed on Ubuntu.

Thanks again!

rmg commented:
From the docs docs for [update()](https://docs.mongodb.org/manual/reference/method/db.collection.update/#db.collection.update):
> Changed in version 2.6: The update() method returns an object that contains the status of the operation.

Amir-61 commented:
>From the docs docs for update():
>Changed in version 2.6: The update() method returns an object that contains the status of the operation.

Oh I see thanks! @bajtos had guessed that could be due different versions of the MongoDB server as well. I just installed version of 2.4.14 Mongodb Server and with this version 11 test cases fail; good to know that my patch is not compatible with older versions...

Thanks!

rmg commented:
If we decided that compatibility with MongoDB older than 2.6 isn't important then we can work on updating the CI environment for it. I don't know what the usage is like, though.

bajtos commented:
I think dropping compatibility with MongoDB older than 2.6 is a breaking change, which I would rather avoid at this time.

> > From the docs docs for update():
> > Changed in version 2.6: The update() method returns an object that contains the status of the operation.

> Oh I see. Thanks! @bajtos had guessed that could be due different versions of the MongoDB server as well. I just installed version of 2.4.14 Mongodb Server and with this version 11 (instead of 6) test cases fail; good to know that my patch is not compatible with older versions...

This isn't new, I had to deal with different MongoDB versions in #121. @Amir-61 please take a closer look at what are the responses returned by different MongoDB versions, I hope we will be able to find a solution that works on both versions. In the worst case, we can return `isNewInstance:undefined` for the older version.

Amir-61 commented:
I tried mongodb version `2.6.11` on my local machine and interestingly all test cases passed; whereas 11 test cases fail on mongodb version `2.4.14`; @rmg could you please confirm what exact version of mongodb you are using on Ubuntu servers? That would help a lot!

CC: @bajtos I may not get enough time to fix this issue in this sprint; thus, I created a [task](https://github.com/strongloop-internal/scrum-loopback/issues/795) for this and moved it to my TOB to take care of it ASAP.

Thanks!
Amir

rmg commented:
@Amir-61 CI is running mongodb v2.4.9 on Ubuntu - the version is listed in the build console log.

I don't understand why you are asking, though. You appear to have successfully reproduced the problem and confirmed that you can make your tests fail locally when using mongodb 2.4.. Isn't that all you need to be able to work on a solution that works for both 2.4 and 2.6?

Amir-61 commented:
> Isn't that all you need to be able to work on a solution that works for both 2.4 and 2.6?
 Oh my bad I thought Ubuntu is using 2.6 and it is failing; Yes I'm good then and could reproduce the issue successfully. Thanks a lot! :pray:

Amir-61 commented:
@slnode test please

bajtos commented:
@slnode test please

Amir-61 commented:
Thi [PR](https://github.com/strongloop/loopback-datasource-juggler/pull/866) needs to be landed before this patch in `loopback-connector-mongodb`.

Amir-61 commented:
@bajtos PTAL; Please note, I did not squash the commits into one single commit, so that you can have an easier review.

Thanks again!

bajtos commented on test/persistence-hooks.test.js:
```diff
@@ -15,4 +15,6 @@ for(var i in config) {
-suite(global.getDataSource(customConfig), should);
+suite(global.getDataSource(customConfig), should,
+  {replaceOrCreateReportsNewInstance: (process.env.MONGODB_VERSION &&
+  !semver.satisfies(process.env.MONGODB_VERSION, '>=2.6.0'))});
```
This is too terse to my taste. Could you please refactor so that we can end up with something like the following?

```js
suite(global.getDataSource(customConfig), should, {
  replaceOrCreateReportsNewInstance: /* a short statement */
});
```

For example:

```js
var DB_VERSION = process.env.MONGODB_VERSION;
var DB_HAS_2_6_FEATURES = semver.satisfies(DB_VERSION, '>=2.6.0');

suite(global.getDataSource(customConfig), should, {
  replaceOrCreateReportsNewInstance: DB_HAS_2_6_FEATURES
});
```

bajtos commented:
@Amir-61 nice! One more minor comment, see above.

bajtos commented:
@Amir-61 I think it's ok to squash all commits into a single one while fixing my last comment, so that we can save one round-trip and get this landed sooner.

Amir-61 commented on test/persistence-hooks.test.js:
```diff
@@ -14,5 +14,10 @@ if (process.env.MONGODB_VERSION &&
   customConfig[i] = config[i];
 }
+var DB_VERSION = process.env.MONGODB_VERSION;
+var DB_HAS_2_6_FEATURES = (!DB_VERSION ||
```
@bajtos: Please note: if you do `npm test` locally we assume you have mongodb 2.6+ and supports `isNewInstance`

bajtos replied:
Could you please add a `console.log` to let the users know about this fact?

For example:

```js
if (!DB_VERSION) {
  console.log('The ENV variable MONGODB_VERSION is not set. Assuming MongoDB version 2.6 or newer.');
}
```

bajtos commented:
The last nitpick: the commit message says `Implementation for replaceAttribute`, but I believe the connector method is called `replaceById`, isn't it?

bajtos commented:
@Amir-61 I left two minor comments to address, no further reviews are necessary. Please get the CI builds green before landing this patch. I guess the juggler PR needs to get landed first?

Amir-61 commented:
@slnode test please
